framework/decode/vulkan_rebind_allocator.cpp

```diff
@@ -899,6 +939,187 @@ void VulkanRebindAllocator::ReportBindImage2Incompatibility(uint32_t
+        }
+    }
+}
+
+void VulkanRebindAllocator::WriteBoundResourceStaging(
+    ResourceAllocInfo* resource_alloc_info, size_t src_offset, size_t dst_offset, size_t data_size, const uint8_t* data)
+{
+    if (resource_alloc_info->is_image && dst_offset != 0)
+    {
+        // TODO implement offset based stagging image copy
+        return;
+    }
+
+    VkBuffer           staging_buf{};
+    VkBufferCreateInfo staging_buf_create_info{};
+    staging_buf_create_info.sType = VK_STRUCTURE_TYPE_BUFFER_CREATE_INFO;
+    staging_buf_create_info.size  = data_size;
+    staging_buf_create_info.usage = VK_BUFFER_USAGE_TRANSFER_SRC_BIT;
+
+    VmaAllocation           staging_alloc{};
+    VmaAllocationInfo       staging_alloc_info{};
+    VmaAllocationCreateInfo staging_alloc_create_info = {};
+    staging_alloc_create_info.flags =
+        VMA_ALLOCATION_CREATE_HOST_ACCESS_SEQUENTIAL_WRITE_BIT | VMA_ALLOCATION_CREATE_MAPPED_BIT;
+    staging_alloc_create_info.usage = VMA_MEMORY_USAGE_CPU_ONLY;
+
+    VkResult result = vmaCreateBuffer(allocator_,
+                                      &staging_buf_create_info,
+                                      &staging_alloc_create_info,
+                                      &staging_buf,
+                                      &staging_alloc,
+                                      &staging_alloc_info);
+
+    void* copy_mapped_pointer{ resource_alloc_info->mapped_pointer };
+
+    if (result == VK_SUCCESS)
+    {
+        result = vmaMapMemory(allocator_, staging_alloc, &resource_alloc_info->mapped_pointer);
+    }
+
+    if (result == VK_SUCCESS)
+    {
+        WriteBoundResourceDirect(resource_alloc_info, src_offset, 0, data_size, data);
+        resource_alloc_info->mapped_pointer = copy_mapped_pointer;
+        vmaFlushAllocation(allocator_, staging_alloc, 0, VK_WHOLE_SIZE);
+        vmaUnmapMemory(allocator_, staging_alloc);
+
+        VkCommandBufferBeginInfo cmd_buf_begin_info = {};
+        cmd_buf_begin_info.sType                    = VK_STRUCTURE_TYPE_COMMAND_BUFFER_BEGIN_INFO;
+
+        result = functions_.begin_command_buffer(cmd_buffer_, &cmd_buf_begin_info);
+    }
+
+    if (result == VK_SUCCESS)
+    {
+        if (resource_alloc_info->is_image)
+        {
+            VkImage original_image{};
+
+            for (const std::pair<const VkImage, VulkanRebindAllocator::ResourceAllocInfo*>& elt :
+                 resource_alloc_info->memory_info->original_images)
+            {
+                if (elt.second == resource_alloc_info)
+                {
+                    original_image = elt.first;
+                    break;
+                }
+            }
+
+            if (original_image)
+            {
+                VkBufferImageCopy region{};
+                region.bufferOffset      = 0;
+                region.bufferRowLength   = 0;
+                region.bufferImageHeight = 0;
+                region.imageSubresource  = { VK_IMAGE_ASPECT_COLOR_BIT, 0, 0, 1 };
```

Comment:
What happens in case of multiple mip maps / array layers?

Reply:
In the current implementation the other subresources would be ignored (except the first one). This is maybe a TODO for later.

Reply:
In that case I believe a `GFXRECON_LOG_WARNING` would be necessary, otherwise you might still get corruption (although less likely) but  you won't get any hints about what might be causing it

Reply:
Thank you for the suggestion, i have added the warning messages